### PR TITLE
coordinator: auto mode is supported

### PR DESCRIPTION
**Pull request by cyclinder**
---
name: Pull request
about: Tell us about your contribution
labels: ["pr/release/none-required"]
---

---
## Thanks for contributing !

Before submitting a pull request, make sure you read about our Contribution notice here: <https://spidernet-io.github.io/spiderpool/contributing/development/pullrequest/#need-review>

**What this PR does / why we need it**:

We do not need to manually specify the mode in the CNI configuration file, which can be determined automatically by the code.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**make sure your commit is signed off**

### CODE REVIEW

File: cmd/coordinator/cmd/command_add.go [hunk 1]
```diff
@@ -110,14 +112,14 @@ func CmdAdd(args *skel.CmdArgs) (err error) {
...
+	err = c.coordinatorModeAndFirstInvoke(conf.PodDefaultCniNic)
 	if err != nil {
 		logger.Error(err.Error())
 		return err
 	}
 
 	// get basic info
-	switch conf.Mode {
+	switch c.tuneMode {
 	case ModeUnderlay:
```
Comment (weizhoublue) [inline cmd/coordinator/cmd/command_add.go:123]: for the switch , no case for auto ?

File: cmd/coordinator/cmd/utils.go [hunk 2]
```diff
@@ -26,20 +26,55 @@ type coordinator struct {
...
 	hostVethName, podVethName, currentInterface, interfacePrefix string
-	HijackCIDR                                                   []string
+	HijackCIDR, podNics                                          []string
 	netns                                                        ns.NetNS
 	hostVethHwAddress, podVethHwAddress                          net.HardwareAddr
 	currentAddress                                               []netlink.Addr
 	hostIPRouteForPod                                            []net.IP
 }
 
+func (c *coordinator) autoMode(mode Mode, podFirstInterface string) error {
```
Comment (weizhoublue) [inline cmd/coordinator/cmd/utils.go:36]: func (c *coordinator)  autoModeToSpecificMode(

File: cmd/coordinator/cmd/utils.go [hunk 3]
```diff
@@ -26,20 +26,55 @@ type coordinator struct {
...
+	if vethExist {
+		c.tuneMode = ModeUnderlay
+	} else {
+		c.tuneMode = ModeOverlay
+		// If spinderpool only assigns a NIC to the pod, Indicates that it is the first invoke
+		if len(c.podNics) == 1 {
+			c.firstInvoke = true
+		}
+	}
+
```
Comment (weizhoublue) [inline cmd/coordinator/cmd/utils.go:62]: add some log about changing auto to xxx mode ?

File: docs/usage/coordinator-zh_CN.md [hunk 4]
```diff
@@ -71,6 +71,8 @@ spec:
 
 - mode: 指定 coordinator 运行在 underlay 模式
 
+> 在默认情况下 mode 的值为auto, 这意味着 coordinator 将会自动判断 mode 为 underlay. 您只需要在 Pod 注入注解: `v1.multus-cni.io/default-network: kube-system/macvlan-underlay`。
```
Comment (weizhoublue) [inline docs/usage/coordinator-zh_CN.md:74]: 得加一些 auto 判断 为 underlay 和 overlay 的 具体算法，后续可能会作为排障的 review 依据

File: docs/usage/coordinator-zh_CN.md [hunk 5]
```diff
@@ -69,7 +72,7 @@ spec:
     }
 ```
 
-- mode: 指定 coordinator 运行在 underlay 模式
+- mode: 指定 coordinator 运行在 underlay 模式。或默认为 auto 模式，您只需要在 Pod 注入注解: `v1.multus-cni.io/default-network: kube-system/macvlan-underlay`, coordinator 将会自动判断 mode 为 underlay。
```
Comment (weizhoublue) [inline docs/usage/coordinator-zh_CN.md:75]: 您只需要在 Pod 注入注解: `v1.multus-cni.io/default-network: kube-system/macvlan-underlay`, coordinator 将会自动判断 mode 为 underlay

---

不是这个逻辑吧？anyway，是否有一个代码的文字逻辑说明，告诉排障人员这个判断的逻辑，而不是每次需要看代码

Comment (cyclinder) [reply docs/usage/coordinator-zh_CN.md:75]: 在 26 行

Comment (cyclinder) [reply docs/usage/coordinator-zh_CN.md:75]: 不过应该也漏了句话

File: docs/usage/coordinator-zh_CN.md [hunk 6]
```diff
@@ -23,6 +23,9 @@ ClusterIP 的路由，导致无法访问。
 
 ### 配置 coordinator 运行在 underlay 模式
 
+> 在默认情况下 mode 的值为auto, coordinator 将通过对比当前 CNI 网卡是否是 `eth0`, 如果是，则自动判断为 Underlay 模式。
```
Comment (weizhoublue) [inline docs/usage/coordinator-zh_CN.md:26]: （1）在默认情况下 mode  
----- 在默认情况下，coordinator  CR 的  spec.mode 
（2）auto 模式下，是 如何自动判别的？按照这句话的解释，逻辑不通：
eth0时，coordinator 永远工作在 underlay 模式 ？

Comment (cyclinder) [reply docs/usage/coordinator-zh_CN.md]: Updated.

**Pull request merged**